diesel/src/query_builder/select_by_statement/mod.rs

```diff
@@ -0,0 +1,100 @@
+use super::{AstPass, QueryFragment};
+use crate::backend::Backend;
+use crate::connection::Connection;
+use crate::deserialize::{QueryableByName, TableQueryable};
+// use crate::expression::subselect::ValidSubselect;
+use crate::expression::*;
+use crate::query_builder::{QueryId, SelectQuery};
+use crate::query_dsl::LoadQuery;
+// use crate::query_source::joins::{AppendSelection, Inner, Join};
+use crate::query_source::*;
+use crate::result::QueryResult;
+
+mod dsl_impls;
+
+/// it is like `SelectStatement` but without
+/// `Query`, `ValidSubselect` and `AppendSelection`
+#[derive(Debug, Clone, Copy)]
+#[doc(hidden)]
+#[must_use = "Queries are only executed when calling `load`, `get_result` or similar."]
+pub struct SelectByStatement<Selection, Statement> {
+    pub(crate) select: std::marker::PhantomData<Selection>,
+    pub(crate) inner: Statement,
+}
+
+impl<S, STMT> QueryId for SelectByStatement<S, STMT>
+where
+    STMT: QueryId,
+{
+    type QueryId = SelectByStatement<(), STMT::QueryId>;
+    const HAS_STATIC_QUERY_ID: bool = STMT::HAS_STATIC_QUERY_ID;
+}
```

Comment:
Is this is literally the same sql query as `STMT` we should just forward both to the impls of `STMT`, so something like this should be OK:
```suggestion
impl<S, STMT> QueryId for SelectByStatement<S, STMT>
where
    STMT: QueryId,
{
    type QueryId = STMT::QueryID;
    const HAS_STATIC_QUERY_ID: bool = STMT::HAS_STATIC_QUERY_ID;
}
```

To explain that a bit more: `QueryId` is used for the internal stmt caching mechanism. For static queries in some cases a static compile time id is used as stmt cache key. As `STMT` and `SelectByStatemet<S, STMT>` are the same query at SQL level they can share the same key in the statement cache. This allows to use the already cashed statement if somecone constructs the query normally via a explicit `select` in one case, and with a `select_by` in the other case. As the generate sql + statement is the same, this is a quite nice optimization in my opinion. 

(We should have probably something along this lines as comment here, to explain why this makes sense.)

Reply:
Thanks for kindly explanation!